### DISCUSSION TITLE
Download artifact api endpoint is broken

### DISCUSSION THREAD

**Discussion by luantak** [community/community#88698] [Apps, API and Webhooks]
### Select Topic Area

Bug

### Body

I would like to download an artifact from a github action workflow run.

```ts
                   // this works, just to show that I'm authorized and have access to the artifact
                    const respa = await octokit.rest.actions.getArtifact({
                        owner: wr.repository.owner.login,
                        repo: wr.repository.name,
                        artifact_id: artifact.id,
                    })
                    
                    // this throws because of a 403 response
                    const resp = await octokit.rest.actions.downloadArtifact({
                        owner: wr.repository.owner.login,
                        repo: wr.repository.name,
                        artifact_id: artifact.id,
                        archive_format: "zip"
                    })

                   console.log("doesn't reach here, throws instead because of 403 response")

```
I get the following `error.response` in my catch block:
```ts
response: {
    url: 'https://productionresultssa8.blob.core.windows.net/actions-results/[redacted]/workflow-job-run-[redacted]artifacts/[redacted].zip?[redacted parameters]',
    status: 403,
    headers: {
      'access-control-allow-origin': '*',
      'access-control-expose-headers': 'Content-Length,Content-Type,Date,Server,x-ms-request-id',
      'content-length': '321',
      'content-type': 'application/xml',
      date: 'Thu, 18 Jan 2024 13:52:25 GMT',
      server: 'Windows-Azure-Blob/1.0 Microsoft-HTTPAPI/2.0',
      'x-ms-request-id': '[redacted]'
    },
```

The recieved url field is a valid url where the artifact can be downloaded, so in some sense i get the correct response.

However this is still different from the documented behavior [here](https://docs.github.com/en/rest/actions/artifacts?apiVersion=2022-11-28#download-an-artifact)
> Gets a redirect URL to download an archive for a repository. This URL expires after 1 minute. *Look for Location: in the response header to find the URL for the download. *

The decoded `error.response.data` field is:
```xml
<?xml version="1.0" encoding="utf-8"?>
<Error><Code>AuthenticationFailed</Code><Message>Server failed to authenticate the request. Make sure the value of Authorization header is formed correctly including the signature.
RequestId:00c5b85a-201e-00e8-7515-4a0b96000000
Time:2024-01-18T13:52:25.6559347Z</Message></Error>
```

I don't think that I'm doing anything wrong (hence why I get the correct url back), but for some reason the api says I'm not authorized, while providing data for which authorization is required.

I can work around this for now by catching the error in place and then using the url field instead of the header, I think the main problem is the wrong 403 response code, which should be 302 instead.

**Comment by per1234** [comment]
Hi @lu4p. I encountered the same problem. My situation was a bit different because I was making the request using the [`urllib.request`](https://docs.python.org/3/library/urllib.request.html) Python module. Even though that means the specific code adjustment in my application is different from what would be necessary in your application, it might be that the octokit package you are using has the same behavior.

So in case it will be of use to you, I'll add a link to the detailed description of the problem in the pull request where the fix was applied:

https://github.com/arduino/report-size-deltas/pull/83

The error message mentioned in that description:

```text
<Error><Code>InvalidAuthenticationInfo</Code><Message>Authentication information is not given in the correct format. Check the value of Authorization header.
```

is different from the one you are experiencing:

> ```text
> <Error><Code>AuthenticationFailed</Code><Message>Server failed to authenticate the request. Make sure the value of Authorization header is formed correctly including the signature.
> ```

However, I was getting the same error message as you originally, and it only changed to the message mentioned in the error message once I changed the header from `Authorization: token ...` to `Authorization: Bearer ...` ([either format is valid](https://docs.github.com/en/rest/authentication/authenticating-to-the-rest-api?apiVersion=2022-11-28#authenticating-with-a-personal-access-token:~:text=you%20can%20use%20Authorization%3A%20Bearer%20or%20Authorization%3A%20token) in this application so the use of the `Authorization: token ...` format is not a problem).

**Reply by iguessthislldo** [reply]
Thanks for figuring that out! I was confused because I knew the Authorization was correct, but I wouldn't have guessed the redirect didn't want it. I'm using the Pithub library in Perl, but it doesn't give enough control to change how redirect requests are handled. The underlying LWP::UserAgent library (roughly equivalent to `urllib.request`) might, but I ended up [checking if there was a redirect that failed with 403, striping the Authorization, and trying again](https://github.com/OpenDDS/OpenDDS/blob/2b4392c0c35aec4e6d047ad2d6c78f36915f75de/tools/scripts/gitrelease.pl#L510-L522). This will probably work for other HTTP libraries that don't have a clear equivalent to `add_unredirected_header`.

**Comment by ultimaweapon** [comment]
I also having the same issue. Been fighting with this issue for 2 hours then found this discussion.

**Comment by ryanleecode** [comment]
This issue needs to be resolved ASAP as this is absurd behaviour

**Comment by ZerGo0** [comment]
In case someone stumbles up on this. The fix seems to be that you should not send the auth header to the redirected artifact download request.
